lib/src/revset.rs

```diff
@@ -531,27 +531,40 @@ impl<St: ExpressionState<CommitRef = RevsetCommitRef>> RevsetExpression<St> {
 impl UserRevsetExpression {
     /// Resolve a user-provided expression. Symbols will be resolved using the
     /// provided `SymbolResolver`.
-    // TODO: return ResolvedRevsetExpression which can still be combined
     pub fn resolve_user_expression(
         &self,
         repo: &dyn Repo,
         symbol_resolver: &dyn SymbolResolver,
-    ) -> Result<ResolvedExpression, RevsetResolutionError> {
+    ) -> Result<Rc<ResolvedRevsetExpression>, RevsetResolutionError> {
         resolve_symbols(repo, self, symbol_resolver)
-            .map(|expression| resolve_visibility(repo, &expression))
     }
 }
 
 impl ResolvedRevsetExpression {
-    /// Resolve a programmatically created revset expression and evaluate it in
-    /// the repo.
-    // TODO: rename to .evaluate(), and add .evaluate_unoptimized() for
-    // testing/debugging purpose?
-    pub fn evaluate_programmatic<'index>(
+    /// Evaluates this expression after optimizing it.
```

Comment:
nit: "Optimizes and evaluates this expression" is clearer to me. The "after optimizing it" could otherwise be interpreted as a requirement on the input (i.e. that you have to optimize it before calling this function).

Reply:
Updated.